### PR TITLE
CASMINST-5241: Update NCN CFS configurations in stage 0; remove now-redundant stage 5

### PR DESCRIPTION
**Pull request by mharding-hpe**
# Description

I tested the stage 0 changes (which included the updated scripts) on mug and verified that they work.

This PR makes the following changes:

## `apply_csm_configuration.sh` script

1. Adds the `--no-enable` flag. Normally the script creates an updated CSM configuration for all of the NCNs, disables the NCNs in CFS, re-enables them with the new configuration, and waits until they configure successfully. This flag has the script instead not re-enable them (but instead just updates them to use the new configuration, when they are later enabled).
2. Also change where the script keeps its files, so that they can be located and backed up during a CSM upgrade.

### Stage 0

Direct the upgrader to follow the Getting Started Guide in most cases, except in the case where CSM is the only product installed on the system. This is only expected to be the case for internal installs. In that case, we instead direct the user to use the `apply_csm_configuration.sh` script (with the new flag) so that the NCNs will use the csm-1.3 configuration when they run CFS. This will prevent the errors reported in [CASMINST-5241](https://jira-pro.its.hpecorp.net:8443/browse/CASMINST-5241).

### Stage 2

Modify the commands where the artifacts are copied from ncn-m001 to ncn-m002. Instead of copying the log files, now it creates a tar archive of them, and also pulls in the files from the `apply_csm_configuration.sh` script (if any -- it's fine if there are none). This archive is both uploaded into S3 and is copied to ncn-m002 (in case it is needed for more immediate reference).

### Stage 5

This stage (NCN personalization) is now obsolete, between the GSG changes and the stage 0 changes made with this PR. This PR removes the stage. Stage 4 now proceeds to CSM health validation.

### `prerequisites.sh`

I added a step that takes a snapshot of the CFS components and configurations, and the state of the Kubernetes pods, at the start of the upgrade. The CFS stuff is so there is a backup copy, since we modify it during the upgrade. The K8s state is because often we see things and wonder if the problem existed before the upgrade, and cannot remember.

Incidentally, because I modified this script, shell check was run on it, and it identified a couple of legitimate problems in parts of the script that I hadn't modified. This tells me that the PRs for those changes must have been checked in despite the shell check errors. In one case, a nonexistent variable was being used in an rm command, so not a good thing to let slip in.

For the less serious shell check complaints, I added comments to skip the checks, only because of how late we are in 1.3. For my main branch backport of this, I plan to address them properly.

### Stopping typescripts

I addressed a small piece of feedback from a recent PR that didn't get merged at the time. Now when the upgrade steps tell the user to stop the typescript, they always call out that the `exit` command is used to do this.

# Checklist Before Merging

- [X] If I added any command snippets, the steps they belong to follow the prompt conventions (see [example][1]).
- [X] If I added a new directory, I also updated `.github/CODEOWNERS` with the corresponding team in [Cray-HPE][2].
- [X] My commits or Pull-Request Title contain my JIRA information, or I do not have a JIRA.

[1]: https://github.com/Cray-HPE/docs-csm/blob/main/introduction/documentation_conventions.md#using-prompts
[2]: https://github.com/Cray-HPE/teams

### CODE REVIEW

**Comment by mharding-hpe** [comment]
I'm marking this ready for review, but I am also going to add the do not merge tag on it, because I still need to do testing. I just want to make it available for review in the meantime.

**Review by rbak-hpe — APPROVED** [review]

**Review by denniswalker — APPROVED** [review]

**Review by bklei — APPROVED** [review]
just two spelling errors -- jenkins probably also found them..

File: upgrade/scripts/upgrade/prerequisites.sh [hunk 1]
```diff
@@ -67,14 +67,48 @@ if [[ -z ${CSM_ARTI_DIR} ]]; then
...
+    echo "Backing up CFS configurations to ${CFS_CONFIG_SNAPSHOT}"
+    curl -k -H "Authorization: Bearer $TOKEN" https://api-gw-service-nmn.local/apis/cfs/v2/configurations > "${CFS_CONFIG_SNAPSHOT}"
+
+    CFS_COMP_SNAPSHOT="${SNAPSHOT_DIR}/cfs_components.json"
+    echo "Backing up CFS components to ${CFS_COMP_SNAPSHOT}"
+    curl -k -H "Authorization: Bearer $TOKEN" https://api-gw-service-nmn.local/apis/cfs/v2/components > "${CFS_COMP_SNAPSHOT}"
+
+    # Record state of Kubernetes pods. If a pod is later seen in an unexpected state, this can provide a reference to
+    # determine whether or not the issue existed prior to the upgrade.
+    K8S_PODS_SNAPSHOT="${SNAPSHOT_DIR}/k8s_pods.json"
```
Comment (mharding-hpe) [inline upgrade/scripts/upgrade/prerequisites.sh:99]:
```suggestion
    K8S_PODS_SNAPSHOT="${SNAPSHOT_DIR}/k8s_pods.txt"
```

File: scripts/operations/configuration/apply_csm_configuration.sh [hunk 2]
```diff
@@ -239,10 +266,19 @@ if [[ -n ${CLEAR_STATE} ]]; then
...
-for xname in ${XNAME_LIST}; do
-    run_cmd cray cfs components update ${xname} --enabled true --error-count 0 --desired-config ncn-personalization
-done
+if [[ -z ${NO_ENABLE} ]]; then
+    echo "Setting desired configuration, clearing erorr count, and enabling all listed components"
+    for xname in ${XNAME_LIST}; do
+        run_cmd cray cfs components update ${xname} --enabled true --error-count 0 --desired-config ncn-personalization
+    done
+else
+    echo "Setting desired configuration and clearing erorr count for all listed components"
```
Comment (mharding-hpe) [inline scripts/operations/configuration/apply_csm_configuration.sh:275]:
```suggestion
    echo "Setting desired configuration and clearing error count for all listed components"
```

**Comment by mharding-hpe** [comment]
I tested the stage 0 changes (which included the updated scripts) on mug and verified that they work.

**Comment by mharding-hpe** [comment]
This PR is ready for merge. The main branch backport is still undergoing some edits.

**Pull request merged**